pkg/controller/vitesskeyspace/reconcile_handler.go

```diff
@@ -0,0 +1,124 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vitesskeyspace
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/vttablet/tmclient"
+	"vitess.io/vitess/go/vt/wrangler"
+
+	v2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/reconciler"
+	"planetscale.dev/vitess-operator/pkg/operator/toposerver"
+)
+
+// reconcileHandler provides context for this specific reconcile loop,
+// and handles reconcile related subroutines.
+type reconcileHandler struct {
+	client              client.Client
+	recorder            record.EventRecorder
+	reconciler          *reconciler.Reconciler
+	vtk                 *v2.VitessKeyspace
+	oldStatus           *v2.VitessKeyspaceStatus
+	untouchedConditions map[v2.VitessKeyspaceConditionType]bool
+	// This field holds a toposerver connection. Please don't try to access until you have
+	// run tsInit().
+	ts *toposerver.Conn
+
+	// This field holds a Wrangler. Please don't try to access until you have run tsInit()
+	wr *wrangler.Wrangler
+	// This field holds a tablet manager client internally for closing upon collection of reconcileHandler.
+	// Please don't try to access until you have run tsInit().
+	tmc tmclient.TabletManagerClient
+}
+
+// tsInit will initialize a toposerver connection, as well as a
+// tablet manager client and wrangler for subroutine use.
+func (r *reconcileHandler) tsInit(ctx context.Context) error {
+	if r == nil || r.ts != nil {
+		return nil
+	}
+
+	// We need to initialize for the first time if we got here.
+	ts, err := toposerver.Open(ctx, r.vtk.Spec.GlobalLockserver)
+	if err != nil {
+		r.recorder.Eventf(r.vtk, v1.EventTypeWarning, "TopoConnectFailed", "failed to connect to global lockserver: %v", err)
+		// Give the lockserver some time to come up.
+		log.Info("Could not connect to topo at vitesskeyspace controller.")
+		return err
+	}
+	r.ts = ts
+
+	if r.tmc == nil {
+		r.tmc = tmclient.NewTabletManagerClient()
+	}
+
+	// Wrangler wraps the necessary clients and implements
+	// multi-step Vitess cluster management workflows.
+	wr := wrangler.New(logutil.NewConsoleLogger(), r.ts.Server, r.tmc)
+	r.wr = wr
+
+	return nil
+}
+
+// close should be called in a defer upon construction of a reconcileHandler to
+// defer the closing of underlying topo if we successfully created one.
+func (r *reconcileHandler) close() {
+	if r == nil {
```

Comment:
My thought more here was that if receiver is nil then there's nothing to close. It's true though that r should never really be nil. a defer close call should be done after constructing a reconcileHandler.

Reply:
Maybe r somehow got dropped from memory entirely before close() could get called. Is that possible since close only gets called from a defer? There would be no references left to `r` for the gc to worry about? Or do ref counts still consider refs inside defer calls?

Reply:
Tentatively removed but still curious about this question ^

Reply:
> My thought more here was that if receiver is nil then there's nothing to close. It's true though that r should never really be nil.

In general, I think silently tolerating a nil receiver should be the exception rather than the rule. For most types, we should assume by default that it's a bug to try to call methods on a nil receiver, and letting such calls panic is the best way to catch those bugs.

The exception comes when calling a method on a nil receiver is a useful thing to do on purpose. Some examples could be if the receiver is a map or slice type (since it's conventional to treat nil the same as an empty container), or if the method does some conversion and the correct conversion of a nil value is another nil value.

In this case, if we have a mismatched close() call, I think we'd consider that a mistake and we'd want to be notified about the bug.

> Maybe r somehow got dropped from memory entirely before close() could get called. Is that possible since close only gets called from a defer? There would be no references left to r for the gc to worry about? Or do ref counts still consider refs inside defer calls?

The `defer` statement saves a local copy of any values computed in the defer expression, so they won't be garbage-collected. You can think of `defer r.f(a+g(), b)` as being equivalent to this:

```go
savedFunc := r.f
savedArg1 := a+g()
savedArg2 := b
defer savedFunc(savedArg1, savedArg2)
```

In addition, deferred functions execute before the surrounding function actually returns, so references in function-local variables are still alive:

```go
r := newThing()
defer func() {
    // r is still in scope here so it can't be GC'd
  }()
```